### PR TITLE
Send spec

### PR DESCRIPTION
**Pull request by rakshith91**
Fixes https://github.com/Azure/azure-sdk-for-python/issues/12958

### CODE REVIEW

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 1]
```diff
@@ -36,22 +46,26 @@ def __init__(self, topic_hostname, credential, **kwargs):
         self._topic_hostname = topic_hostname
         auth_policy = _get_authentication_policy(credential)
         self._client = EventGridPublisherClientImpl(authentication_policy=auth_policy, **kwargs)
-
+
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:49]:
```suggestion

```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 2]
```diff
@@ -36,22 +46,26 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
          """
+        if not isinstance(events, list):
+            events = [events]
 
-        if all(isinstance(e, CloudEvent) for e in events):
+        if all(isinstance(e, CloudEvent) for e in events) or all(_is_cloud_event(e) for e in events):
+            kwargs.setdefault("content_type", "application/cloudevents-batch+json; charset=utf-8")
             self._client.publish_cloud_event_events(self._topic_hostname, events, **kwargs)
-        elif all(isinstance(e, EventGridEvent) for e in events):
+        elif all(isinstance(e, EventGridEvent) for e in events) or all(isinstance(e, dict) for e in events):
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:64]: intentionally not doing a client side validation for dicts that are not cloud event

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 3]
```diff
@@ -36,22 +46,26 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
+            kwargs.setdefault("content_type", "application/json; charset=utf-8")
             self._client.publish_events(self._topic_hostname, events, **kwargs)
         elif all(isinstance(e, CustomEvent) for e in events):
             serialized_events = [dict(e) for e in events]
             self._client.publish_custom_event_events(self._topic_hostname, serialized_events, **kwargs)
         else:
-            raise Exception("Event schema is not correct. Please send as list of all CloudEvents, list of all EventGridEvents, or list of all CustomEvents.")
+            raise Exception("Event schema is not correct.")
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:71]:
```suggestion
            raise ValueError("Event schema is not correct.")
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 4]
```diff
@@ -35,21 +47,25 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
+        elif all(isinstance(e, EventGridEvent) for e in events) or all(isinstance(e, dict) for e in events):
+            kwargs.setdefault("content_type", "application/json; charset=utf-8")
+            self._client.publish_events(self._topic_hostname, events, **kwargs)
         elif all(isinstance(e, CustomEvent) for e in events):
             serialized_events = [dict(e) for e in events]
-            await self._client.publish_custom_event_events(self._topic_hostname, serialized_events, **kwargs)
+            self._client.publish_custom_event_events(self._topic_hostname, serialized_events, **kwargs)
         else:
-            raise Exception("Event schema is not correct. Please send as list of all CloudEvents, list of all EventGridEvents, or list of all CustomEvents.")
+            raise Exception("Event schema is not correct.")
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:70]:
```suggestion
            raise ValueError("Event schema is not correct.")
```

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py [hunk 5]
```diff
@@ -108,4 +157,32 @@ def test_eg_publisher_client_publish_custom_schema_event(self, resource_group, e
...
+                    {
+                    "customSubject": "sample2",
+                    "customEventType": "sample.event",
+                    "customDataVersion": "2.0",
+                    "customId": "12345",
+                    "customEventTime": dt.datetime.now(UTC()).isoformat(),
+                    "customData": "sample data 2"
+                    }
+                )
+        client.send([custom_event1, custom_event2])
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py:188]:
```suggestion
        client.send([custom_event1, custom_event2])

```

File: sdk/eventgrid/azure-eventgrid/tests/recordings/test_eg_publisher_client.test_eg_publisher_client_publish_event_grid_event_data_dict.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/tests/recordings/test_eg_publisher_client.test_eg_publisher_client_publish_event_grid_event_data_dict.yaml:1]:
```suggestion
```

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py (hunk 5 above)
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py:188]:
```suggestion
        client.send([custom_event1, custom_event2])

```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 7]
```diff
@@ -36,20 +46,24 @@ def __init__(self, topic_hostname, credential, **kwargs):
         self._topic_hostname = topic_hostname
         auth_policy = _get_authentication_policy(credential)
         self._client = EventGridPublisherClientImpl(authentication_policy=auth_policy, **kwargs)
-
+
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:49]:
```suggestion
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 8]
```diff
@@ -70,3 +70,11 @@ def _get_authentication_policy(credential):
...
         authentication_policy = EventGridSharedAccessSignatureCredentialPolicy(credential=credential, name=constants.EVENTGRID_TOKEN_HEADER)
     return authentication_policy
+
+def _is_cloud_event(event):
+    # type: dict -> bool
+    required = ('id', 'source', 'specversion', 'type')
+    try:
+        return all([_ in event for _ in required]) and event['specversion'] == "1.0"
+    except TypeError:
+        pass
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:80]:
```suggestion
        return False
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 9]
```diff
@@ -35,22 +47,26 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
+            events = [events]
 
-        if all(isinstance(e, CloudEvent) for e in events):
-            await self._client.publish_cloud_event_events(self._topic_hostname, events, **kwargs)
-        elif all(isinstance(e, EventGridEvent) for e in events):
-            await self._client.publish_events(self._topic_hostname, events, **kwargs)
+        if all(isinstance(e, CloudEvent) for e in events) or all(_is_cloud_event(e) for e in events):
+            kwargs.setdefault("content_type", "application/cloudevents-batch+json; charset=utf-8")
+            self._client.publish_cloud_event_events(self._topic_hostname, events, **kwargs)
+        elif all(isinstance(e, EventGridEvent) for e in events) or all(isinstance(e, dict) for e in events):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:64]: So is the semantic that we're asserting that if a dict isn't a cloudevent it must be an egevent?  there's no fallback to custom event, or way to infer that it's an egevent we need to check?

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 10]
```diff
@@ -35,22 +47,26 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
+            kwargs.setdefault("content_type", "application/json; charset=utf-8")
+            self._client.publish_events(self._topic_hostname, events, **kwargs)
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:66]: Nit: seeing this alongside the other publish calls, makes me think if we're specific with those (*_cloud_event_*, *_custom_event*, etc) we should be specific with this too (publish_event_grid_events?)

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py [hunk 11]
```diff
@@ -86,7 +86,6 @@ def remove_resource(self, name, **kwargs):
         if self.is_live:
             group = self._get_resource_group(**kwargs)
             self.client.topics.delete(group.name, name, polling=False)
-
     def _get_resource_group(self, **kwargs):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:89]: seems like we'd want this whitespace?
```suggestion

    def _get_resource_group(self, **kwargs):
```

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py [hunk 12]
```diff
@@ -75,11 +109,26 @@ def test_eg_publisher_client_publish_cloud_event_data_str(self, resource_group,
                 type="Sample.Cloud.Event"
                 )
         client.send([cloud_event])
-    
+
+    @pytest.mark.liveTest
+    @CachedResourceGroupPreparer(name_prefix='eventgridtest')
+    @CachedEventGridTopicPreparer(name_prefix='cloudeventgridtest')
+    def test_send_cloud_event_dict(self, resource_group, eventgrid_topic, eventgrid_topic_primary_key, eventgrid_topic_endpoint):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py:116]: if we're changing to normalize as "test_eg_publisher_client_publish" let's adjust this accordingly.

(Truthfully I think we could have kept it terse with test_publish_*, e.g. test_publish_cloud_event_dict)

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py (hunk 10 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:66]: im open to it - but we would need to change the swagger

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py (hunk 11 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:89]: prolly crept in when i rebased

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 13]
```diff
@@ -36,20 +46,23 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
         :type events: Union[List[models.CloudEvent], List[models.EventGridEvent], List[models.CustomEvent]]
         :rtype: None
         raise: :class:`ValueError`, when events do not follow specified SendType.
          """
+        if not isinstance(events, list):
+            events = [events]
 
-        if all(isinstance(e, CloudEvent) for e in events):
+        if all(isinstance(e, CloudEvent) for e in events) or all(_is_cloud_event(e) for e in events):
+            kwargs.setdefault("content_type", "application/cloudevents-batch+json; charset=utf-8")
```
[the remark this answers is not in the document]
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:63]: fixed

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 14]
```diff
@@ -36,20 +46,24 @@ def __init__(self, topic_hostname, credential, **kwargs):
...
         self._client = EventGridPublisherClientImpl(authentication_policy=auth_policy, **kwargs)
-
     def send(self, events, **kwargs):
-        # type: (Union[List[CloudEvent], List[EventGridEvent], List[CustomEvent]], Any) -> None
+        # type: (SendType, Any) -> None
         """Sends event data to topic hostname specified during client initialization.
 
         :param events: A list of CloudEvent/EventGridEvent/CustomEvent to be sent.
         :type events: Union[List[models.CloudEvent], List[models.EventGridEvent], List[models.CustomEvent]]
+        :keyword str content_type: The type of content to be used to send the events.
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:55]:
```suggestion
        :keyword str content_type: The type of content to be used to send the events.
        Has default value "application/cloudevents-batch+json; charset=utf-8"
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:63]: See above addendum (default)

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 15]
```diff
@@ -6,18 +6,30 @@
...
+    # pylint: disable=unused-import,ungrouped-imports
+    from typing import Any, Union, Dict, List
+    SendType = Union[
+        CloudEvent,
+        EventGridEvent,
+        CustomEvent,
+        Dict,
+        List[CloudEvent],
+        List[EventGridEvent],
+        List[CustomEvent]
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:31]:
```suggestion
        List[CustomEvent],
        List[Dict]
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 16]
```diff
@@ -13,13 +13,23 @@
...
-    from typing import Any
+    from typing import Any, Union, Dict, List
+    SendType = Union[
+        CloudEvent,
+        EventGridEvent,
+        CustomEvent,
+        Dict,
+        List[CloudEvent],
+        List[EventGridEvent],
+        List[CustomEvent]
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:24]:
```suggestion
        List[CustomEvent],
        List[Dict]
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 17]
```diff
@@ -30,23 +42,29 @@ class EventGridPublisherClient(object):
...
 
 
     async def send(self, events, **kwargs):
-        # type: (Union[List[CloudEvent], List[EventGridEvent], List[CustomEvent]], Any) -> None
+        # type: (SendType) -> None
         """Sends event data to topic hostname specified during client initialization.
 
         :param  events: A list of CloudEvent/EventGridEvent/CustomEvent to be sent.
         :type events: Union[List[models.CloudEvent], List[models.EventGridEvent], List[models.CustomEvent]]
+        :keyword str content_type: The type of content to be used to send the events.
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:56]:
```suggestion
        :keyword str content_type: The type of content to be used to send the events.
        Has default value "application/cloudevents-batch+json; charset=utf-8"
```

**Pull request merged**